data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -231,7 +231,24 @@ public void nonNullWrite(int rowId, String data, ColumnVector output) {
 
     @Override
     public void nonNullWrite(int rowId, ByteBuffer data, ColumnVector output) {
-      ((BytesColumnVector) output).setRef(rowId, data.array(), 0, data.array().length);
+      // We technically can't be sure if the ByteBuffer coming in is on or off
+      // heap so we cannot safely call `.array()` on it without first checking
+      // via the method ByteBuffer.hasArray().
+      // See: https://errorprone.info/bugpattern/ByteBufferBackingArray
+      //
+      // When there is a backing heap based byte array, we avoided the overhead of
+      // copying, which is especially important for small byte buffers.
+      //
+      // TODO - This copy slows it down, perhap unnecessarily. Is there any other way to tell, or no?
+      //        My guess is no, if I consider things like VectorizedOrcReaders on Spark.
+      if (data.hasArray()) {
+        ((BytesColumnVector) output).setRef(rowId, data.array(), 0, data.array().length);
```

Comment:
This is not a correct use of `ByteBuffer` because it doesn't use `arrayOffset` or `remaining`. I think the current version must work because Spark returns new arrays that we wrap in `ByteBuffer`, but if the goal here is to make this accept any `ByteBuffer` then we should account for cases where the buffer is not simply the entire backing array.

Reply:
I'd like to point out that this is the original version of `nonNullWrite`, unless you're referring to the `else` block lines 246-251 just below your comment, in which case that's my addition.  https://github.com/apache/iceberg/pull/1334/files#diff-b1b07b15f036000a3f2bed76fdd9f961R246-R251

I believe my addition (in the else block) is correct as it does call `remaining`. If we use `hasArray`, I just left it as the original function. I'd be happy to update anything that needs updating though.

Reply:
Ok I see what you're referring to now. I will handle the fix in this PR as well. 👍

Reply:
Should we consume the buffer up until that point or reset it to its original position?

Reply:
If we are accessing the backing array, there is no need to worry about the state of the `ByteBuffer`. But the actual starting offset in the array is `data.arrayOffset() + data.position()` and the length is `data.remaining()`. Have a look at the copy methods in our `ByteBuffers` class to see examples.